router/src/lib.rs

```diff
@@ -520,6 +522,126 @@ pub(crate) struct ChatRequest {
+    #[serde(default)]
+    #[schema(nullable = true, example = "null")]
+    #[serde(deserialize_with = "deserialize_tool_choice::deserialize")]
+    pub tool_choice: Option<ToolType>,
+}
+
+fn default_tool_prompt() -> Option<String> {
+    Some(
+        "\nBased on the conversation, please choose the most appropriate tool to use: ".to_string(),
+    )
+}
+#[derive(Clone, Deserialize, ToSchema, Serialize)]
+enum ToolType {
+    FunctionName(String),
+    OneOf,
+}
+
+/// Deserialize the tool choice from the JSON input or from the function name ("none" is allowed but mapped to None)
+mod deserialize_tool_choice {
+    use super::*;
+    use serde::de;
+    use serde::Deserializer;
+    use serde_json::Value;
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<ToolType>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value = Value::deserialize(deserializer)?;
+
+        match value {
+            Value::String(s) => match s.as_str() {
+                "none" => Ok(None),
+                "auto" => Ok(Some(ToolType::OneOf)),
+                _ => Ok(Some(ToolType::FunctionName(s))),
+            },
+            Value::Object(map) => {
+                if let Some(content) = map
+                    .get("function")
+                    .and_then(|v| v.get("name"))
+                    .and_then(|v| v.as_str())
+                {
+                    Ok(Some(ToolType::FunctionName(content.to_string())))
+                } else {
+                    Err(de::Error::custom("function key not found in tool choice"))
+                }
+            }
+            Value::Null => Ok(Some(ToolType::OneOf)),
+            _ => Err(de::Error::custom("invalid token format")),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, ToSchema, Default)]
+pub struct Tools {
+    // rename to "$function" to avoid conflicts with other fields
+    #[serde(rename = "$function")]
+    pub function: std::collections::HashMap<String, serde_json::Value>,
+    pub any_of: Vec<FunctionRef>,
+}
+
+// Allows Tools to be converted to a valid JSON schema object
+impl From<Tools> for serde_json::Value {
+    fn from(tools: Tools) -> Self {
+        let mut map = serde_json::Map::new();
+        let mut functions = serde_json::Map::new();
+        for (name, value) in tools.function {
+            functions.insert(name, value);
+        }
+        map.insert("$functions".to_string(), serde_json::json!(functions));
+        let mut properties = serde_json::Map::new();
+        let mut function = serde_json::Map::new();
+        function.insert("anyOf".to_string(), serde_json::json!(tools.any_of));
+        properties.insert("function".to_string(), serde_json::json!(function));
+        map.insert("properties".to_string(), serde_json::json!(properties));
+        serde_json::Value::Object(map)
+    }
+}
```

Comment:
This should be handled by serde directly.

`Serde(rename="camelCase")` probably ?)

Reply:
this logic is a bit different than the default serde logic and builds a nested json object from the `Vec<FunctionRef>` but I think it can be simplified or removed in favor of serde builtins (will update soon!) thanks

Reply:
You can also use intermediary structs to have "raw" format, and pass on to internal structs handled by us (to remove redundancies, etc..)